backend-tests/tests/test_deployments.py

```diff
@@ -0,0 +1,387 @@
+            qs_params={"artifact_name": artifact_name,
+                       "device_type"  : "qemux86-64"}
+        )
+        assert resp.status_code == expected_status_code
+        if resp.status_code == 200:
+            # Update device status upon successful request
+            api_dev_deploy.with_auth(dev.token).call(
+                "PUT",
+                deployments.URL_STATUS.format(id=resp.json()["id"]),
+                body={"status": "success"}
+            )
+
+
+
+class TestDeploymentsBackend(object):
+    def test_regular_deployment(self, setup_deployments_enterprise_test):
+        user, tenant, utoken, devs = setup_deployments_enterprise_test
+
+        api_mgmt_dep = ApiClient(deployments.URL_MGMT)
+
+        # Make deployment request
+        deployment_req = {
+            "name"          : "phased-deployment",
+            "artifact_name" : "deployments-phase-testing",
+            "devices"       : [dev.id for dev in devs]
+        }
+        api_mgmt_dep.with_auth(utoken).call(
+            "POST",
+            deployments.URL_DEPLOYMENTS,
+            deployment_req
+        )
+
+        try_update(
+            devs,
+            expected_status_code=200,
+            artifact_name="bugs-bunny"
+        )
+        try_update(
+            devs,
+            expected_status_code=204,
+            artifact_name="bugs-bunny"
+        )
+
+        deployment_req["name"] = "really-old-update"
+        api_mgmt_dep.with_auth(utoken).call(
+            "POST",
+            deployments.URL_DEPLOYMENTS,
+            deployment_req
+        )
+        try_update(
+            devs,
+            expected_status_code=204,
+            artifact_name=deployment_req["artifact_name"]
+        )
+
+
+class TestPhasedRolloutDeploymentsBackend:
+    def try_phased_updates(self, deployment, devices, user_token):
+        ### Static helper function ###
+        # Setup Deployment APIs
+        api_mgmt_deploy = ApiClient(deployments.URL_MGMT)
+
+        # POST deployment
+        resp = api_mgmt_deploy.with_auth(user_token).call(
+            'POST',
+            deployments.URL_DEPLOYMENTS,
+            body=deployment
+        )
+        assert resp.status_code == 201
+
+        # Store the location from the GET /deployments/{id} request
+        deployment_id = os.path.basename(resp.headers['Location'])
+        devices_provisioned = 0
+        num_phases = len(deployment["phases"])
+
+        for i in range(num_phases):
+            if i == 0 and "start_ts" not in deployment["phases"][i]:
+                # First phase don't need to specify `start_ts`
+                pass
+            elif "start_ts" in deployment["phases"][i]:
+                # Sleep until next phase starts
+                start_ts = datetime.strptime(
+                    deployment["phases"][i]["start_ts"],
+                    "%Y-%m-%dT%H:%M:%SZ"
+                )
+                now = datetime.utcnow()
+                wait_time = start_ts - now
```

Comment:
does this mean we're waiting for each phase to start before ever pinging 'next'?

that's 'good', and must work, but note that one of the most useful checks for phased rollout is:
to make sure that in between phases, devices will not get an update.

I'm not sure we're testing this, it appears not.

remember this is the top business value here - a user can inspect a deployment after phase 1(say 5%), and has a time window to abort the whole thing if it went sideways. so this is the primary thing to test.

so it would be good to test:
- in between phases - that nothing can get through
- while a phase is in progress - ping `GET /deployment/id` to see if the device count is incremented correctly

Reply:
cool, I think you caught everything here:
- no device can get in before a phase starts - check
- devs from prev phase get 204 already-installed, and don't update the phase cnt - check
- phase count incrementation during the phase - check, actually device-by-device
- phase count is correct after the full phase - check
- after a phase is saturated - next devices in line can't get in - check
- deployment is 'finished' after all phases

whew, I can't think of anything more, so that's +1.